src/server/utils/dates.py

```diff
@@ -67,3 +74,82 @@ def weeks_in_range(week_range: Tuple[int, int]) -> int:
         year = Year(y)
         acc += year.totalweeks()
     return acc + 1  # same week should lead to 1 week that will be queried
+
+def dates_to_ranges(values: Optional[Sequence[Union[Tuple[int, int], int]]]) -> Optional[Sequence[Union[Tuple[int, int], int]]]:
+    """
+    Converts a mixed list of dates and date ranges to an optimized list where dates are merged into ranges where possible.
+    e.g. [20200101, 20200102, (20200101, 20200104), 20200106] -> [(20200101, 20200104), 20200106]
+    (the first two values of the original list are merged into a single range)
+    """
+    if not values or len(values) == 0:
+        return values
+
+    # determine whether the list is of days (YYYYMMDD) or weeks (YYYYWW) based on first element
+    try:
+        if (isinstance(values[0], tuple) and guess_time_value_is_day(values[0][0]))\
+            or (isinstance(values[0], int) and guess_time_value_is_day(values[0])):
+            return days_to_ranges(values)
+        else:
+            return weeks_to_ranges(values)
+    except:
+        return values
+
+def days_to_ranges(values: Sequence[Union[Tuple[int, int], int]]) -> Sequence[Union[Tuple[int, int], int]]:
+    dates = []
+
+    # populate list of days
+    for v in values:
+        if isinstance(v, int):
+            dates.append(time_value_to_date(v))
+        else: # tuple
+            start_date = time_value_to_date(v[0])
+            end_date = time_value_to_date(v[1])
+            # add all dates between start and end, inclusive of end date
+            dates.extend([start_date + timedelta(n) for n in range(int((end_date - start_date).days) + 1)])
+
+    # remove duplicates then sort
+    dates = sorted(list(set(dates)))
```

Comment:
issue: this could be potentially slow if someone requests a range like 19000101-20200101 (and there is a place in the [codebase](https://github.com/cmu-delphi/delphi-epidata/blob/ds/jit-update-meta/src/acquisition/covidcast/database_meta.py#L330-L348) where I have to do that right now, because I can't use `time_value=*` in JIT for complicated reasons...).

suggestion: have you thought about a slightly different approach where you convert everything to a range (e.g. 20220101 -> (20220101, 20220101)), sort, and then write some logic to merge overlapping / neighboring ranges? there could probably be an interesting approach with `groupby` there.

if it seems too complicated, I'm ok with filing this as a separate enhancement issue and coming back to it. up to you.

Reply:
That's an excellent suggestion! I implemented your alternate approach and reworking the logic this way lowered the execution time for your edge case (1900-2050) from ~89 ms to 0.1 ms :)